relayer/src/client.rs

```diff
@@ -1,120 +1,284 @@
-                }
+            if Some(finalized_block.number) == latest_fetched_block {
+                log::info!(target: TARGET,"No new finalized blocks, retrying in {} seconds", duration.period().as_secs());
+                continue;
             }
+            log::info!(target: TARGET,"New blocks to fetch. Latest finalized: {}, Latest processed: {latest_fetched_block:?}", finalized_block.number );
+
+            if let Err(e) = self
+                .collect_blocks_after_finality_update(finalized_block, latest_fetched_block)
+                .await
+            {
+                log::error!(target: TARGET,"Failed to process finality update: {}", e);
+            } else {
+                log::info!(target: TARGET,"Processed finality update");
+            };
+
+            // Update latest fetched block after fetching. This is needed to avoid querying db on every iteration.
+            latest_fetched_block = self.db.select_latest_fetched_block_height()?;
         }
     }
-}
 
-fn network(config: &Config) -> Network {
-    match &config.network {
-        Some(network) => {
-            if network == "goerli" {
-                Network::GOERLI
+    /// Fetches all blocks from the web3 provider. The fetching goes backwards from the latest finalized block
+    /// to the latest processed block using parent hash.
+    async fn collect_blocks_after_finality_update(
+        &mut self,
+        finalized_block: ExecutionBlock,
+        latest_fetched_block: Option<u64>,
+    ) -> Result<()> {
+        const TARGET: &str = "relayer::client::collect_blocks_after_finality_update";
+
+        log::info!(target: TARGET,"Processing finality update");
+        let latest_fetched_block =
+            latest_fetched_block.unwrap_or(finalized_block.number - BLOCK_AMOUNT_TO_STORE);
+
+        log::info!(target: TARGET,"Latest fetched block: {}", latest_fetched_block);
+
+        // Now we have fetch missing blocks using previous block hash until we hit latest processed block.
+        // If it's first run, we have to backtrack for BLOCK_AMOUNT_TO_STORE blocks.
+        let mut blocks_to_process =
+            Vec::with_capacity((finalized_block.number - latest_fetched_block) as usize);
+
+        let mut current_block = finalized_block.number - 1;
+        let mut prev_block_hash = finalized_block.parent_hash;
+        let block = self
+            .block_rpc
+            .get_block(finalized_block.hash)
+            .await?
+            .ok_or_else(|| eyre::eyre!("Block not found"))?;
+        // push first finalized block to the queue
+        blocks_to_process.push((parse_block(block)?, H256(finalized_block.hash.0)));
+
+        let mut repeat = 0;
+        let repeat_cycle = |repeat_counter| {
+            const RETRIES: u64 = 10;
+            if repeat_counter < RETRIES {
+                Ok(repeat_counter + 1)
             } else {
-                Network::MAINNET
+                Err(eyre::eyre!("Multiple retries happened"))
```

Comment:
Could you add `log::error!(..)` here?
Perhaps we need to use some context to describe what branch failed.

Reply:
done